### PR TITLE
Fix URL by removing slash and backslash

### LINKED ISSUE #1729: Slash (`/`) in meeting URLs throws a 404
Found a bug? Please fill out the sections below. 👍

### Issue Summary

When editing an event type, you can add a URL with a slash (`/`).  For example, I tried to set up cal.com/anand/hiring/20 for a 20-minute introductory call for hiring. However, this throws a 404 error since the slug only matches one level of route param.

We should either add validation to prevent the use of certain special characters like slash, question mark, and hash, which are reserved for URLs. Alternately, we should allow nested URLs (which would be awesome since I can use /hiring/etc!) and update the route matching for it. In Next.js for example, the file name `[name].jsx` should be replaced with `[...name].jsx` to support this.

CleanShot 2022-02-07 at 23 17 26@2x

### Steps to Reproduce

1. Click on an event to edit it
2. Add a URL with a slash in it
3. Save the event
4. Try to open the event

### PR DESCRIPTION
**Pull request by Jenietoc**
## What does this PR do?

This PR validates that when creating or editing an eventType it is not possible to write `slash` or `backslash`.

Fixes #1729

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- [X] This implementation was tested by creating and editing several newTypes and the validation worked correctly.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code and corrected any misspellings
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: components/eventtype/CreateEventType.tsx [hunk 1]
```diff
@@ -66,8 +66,14 @@ export default function CreateEventTypeButton(props: Props) {
...
-        if (value.title) setValue("slug", value.title.replace(/\s+/g, "-").toLowerCase());
-        else setValue("slug", "");
+        if (value.title) {
+          const newValueWithoutSpace = value.title.replace(/\s+/g, "-");
+          const newValueWithoutSlash = newValueWithoutSpace.replace(/\//g, "");
+          const newValueWithoutBackSlash = newValueWithoutSlash.replace(/\\/g, "");
+          setValue("slug", newValueWithoutBackSlash.toLowerCase());
+        } else {
+          setValue("slug", "");
+        }
```
Comment (zomars) [inline components/eventtype/CreateEventType.tsx:76]: Let's avoid duplicating code by just using our `slugify` function located at `@lib/slugify`

File: pages/event-types/[type].tsx [hunk 2]
```diff
@@ -653,6 +653,12 @@ const EventTypePage = (props: inferSSRProps<typeof getServerSideProps>) => {
                           required
                           className="flex-1 block w-full min-w-0 border-gray-300 rounded-none rounded-r-sm focus:ring-primary-500 focus:border-primary-500 sm:text-sm"
                           defaultValue={eventType.slug}
+                          onKeyUp={(event) => {
+                            const newValueWithoutSpace = String(event.target.value).replace(/\s+/g, "-");
+                            const newValueWithoutSlash = newValueWithoutSpace.replace(/\//g, "");
+                            const newValueWithoutBackSlash = newValueWithoutSlash.replace(/\\/g, "");
+                            formMethods.setValue("slug", newValueWithoutBackSlash.toLowerCase());
+                          }}
```
Comment (zomars) [inline pages/event-types/[type].tsx:661]: Same as above

**Review by zomars — CHANGES_REQUESTED** [review]
Left some comments, we should also avoid unsanitized slugs to reach the DB by updating our schema here:

https://github.com/calcom/cal.com/blob/38b41f0adbdfd25b72510375168cecd2d6f503b9/prisma/zod-utils.ts#L9

To this:

```js
export const eventTypeSlug = z.string().transform((val) => slugify(val.trim()));
```

File: components/eventtype/CreateEventType.tsx (hunk 1 above)
Comment (Jenietoc) [reply components/eventtype/CreateEventType.tsx]: Perfect, I have already implemented slugify

File: pages/event-types/[type].tsx [hunk 3]
```diff
@@ -653,6 +654,9 @@ const EventTypePage = (props: inferSSRProps<typeof getServerSideProps>) => {
                           required
                           className="flex-1 block w-full min-w-0 border-gray-300 rounded-none rounded-r-sm focus:ring-primary-500 focus:border-primary-500 sm:text-sm"
                           defaultValue={eventType.slug}
+                          onKeyUp={(event) => {
+                            formMethods.setValue("slug", slugify(event.target.value));
+                          }}
```
[the remark this answers is not in the document]
Comment (Jenietoc) [reply pages/event-types/[type].tsx:659]: Hi @zomars is this the issue you are talking about?
`formMethods.setValue("slug", slugify(String(event.target.value)));`
Would it look good?

[the remark this answers is not in the document]
Comment (zomars) [reply pages/event-types/[type].tsx:659]: good enough yeah

**Comment by PeerRich** [comment]
hey @Jenietoc, there seems to be conflicts that only the PR creator can fix "This branch has conflicts that must be resolved" 

can you also take a look at @zomars comment: https://github.com/calcom/cal.com/pull/1733#pullrequestreview-876583820?

**Comment by Jenietoc** [comment]
> hey @Jenietoc, there seems to be conflicts that only the PR creator can fix "This branch has conflicts that must be resolved"
> 
> can you also take a look at @zomars comment: [#1733 (review)](https://github.com/calcom/cal.com/pull/1733#pullrequestreview-876583820)?

Hi @PeerRich conflicts have already been resolved

**Comment by PeerRich** [comment]
we moved to turbo repo which caused this branch to fail @Jenietoc. can you take a look at the new file structure? `apps/web`

**Comment by zomars** [comment]
There's seems to be an issue with the classname sorted. We may want to prioritize this #1796

**Comment by Jenietoc** [comment]
Hello  @PeerRich  @zomars , what should I do to fix the errors in this PR? and if you can do merge.

**Comment by PeerRich** [comment]
> Hello  @PeerRich  @zomars , what should I do to fix the errors in this PR? and if you can do merge.

@zomars will review and likely merge today 🙏

**Comment by Jenietoc** [comment]
> > Hello @PeerRich @zomars , what should I do to fix the errors in this PR? and if you can do merge.
> 
> Hey @Jenietoc please rebase/merge from main and solve a conflict with `packages/prisma/zod-utils.ts` I tried to solved it for you but it seems that your [settings doesn't allow me to](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).

I will talk to the person who created the fork to give those permissions, I will also try to solve the conflicts. thank you very much for your time.

**Comment by zomars** [comment]
Merging as it already passed but it got out of date again.

**Pull request merged**